codegen/winterfell/src/air/boundary_constraints.rs

```
@@ -101,7 +104,10 @@ fn add_aux_trace_assertions(func_body: &mut codegen::Function, ir: &Air) {
             split_boundary_constraint(ir.constraint_graph(), constraint.node_index());
         debug_assert_eq!(trace_access.segment, TraceSegmentId::Aux);
 
-        let expr_root_string = expr_root.to_string(ir, ElemType::Ext, TraceSegmentId::Aux);
+        let expr_root_string = match expr_root {
```

Not entirely sure, but I think we should panic here rather than providing a default value. Moreover, zero is invalid for a multi-set boundary.

---

This part handles empty buses constraints that are part of the algebraic graph (inserted in `expand_buses.rs`), for instance:
```
buses {
    multiset p,
    logup q,
}
boundary_constraints {
    enf p.first = null;
    enf q.first = null;
    enf p.last = null;
    enf q.last = null;
}
```

For multiset the constraints are indeed `Sub(p, 1) = 0` so we won't get zero. But for logup they are `Sub(q, 0) = 0`, and in that case, CSE will reduce the `Sub` node to just `q` and the `expr_root` returned by `split_boundary_constraint` will be None.

---

Thanks for the explanation. Can we make the comment a bit clearer though?

---

Sure, I've explained what's happening in 1554db65404a38d85e516cff4e8e7c61add28503, let me know if it's still unclear!